### DISCUSSION TITLE
Reading document metadata from index

### DISCUSSION THREAD

**Discussion by danielwinkler** [ravendb/ravendb#15661] [Q&A]
Dear all,

I wanted to ask if it is possible to include the document metadata into an index.
We use @metadata.Visibility to model a kind of soft delete behavior, so I would like in the query to include something like 

```
    public IRavenQueryable<T> Query<T, TIndexCreator>() where TIndexCreator : AbstractCommonApiForIndexes, new()
        => Session
            .Query<T, TIndexCreator>()
            .Where(e => RavenQuery.Metadata(e)[VisibilityKey] == null || (string?)RavenQuery.Metadata(e)[VisibilityKey] == "Visible")
            ;
```

Potentially I could read and write in the index definition
```
from entity in entities
let @metadata = MetadataFor(entity)
[...]
select new
{
    entity.Prop1,
    @metadata
}
```
but then i will get the error that VisibilityKey is not indexed.

Kind regards,
Daniel

**Comment by garayx** [comment]
Hi,

Yes, you can index the metadata by using the `MetadataFor()` method:

https://github.com/ravendb/ravendb/blob/03905a4077231d55219da422fbac062ca1568a15/test/SlowTests/Core/Utils/Indexes/Companies_CompanyByType.cs#L31

**Comment by danielwinkler** [comment]
Thanks @garayx ,

But in  this case I would have to explicitly extract the property from the metadata and add it to the index.

I would rather like to have the entire @metadata dictionary in the index, so that i can then, when querying the index use again the `RavenQuery.Metadata` and read this properties.
Do you know if this is possible?

**Reply by garayx** [reply]
yes, for the same Index I can get the `metadata` object by simply using the `MetadataFor()` method, 

```
            Map = companies => from company in companies
                               select new
                               {
                                   Type = company.Type,
                                   ContactsCount = company.Contacts.Count,
                                   LastModified = MetadataFor(company).Value<DateTime>(Constants.Documents.Metadata.LastModified),
                                   Metadata = MetadataFor(company)
                               };
```

![image](https://user-images.githubusercontent.com/16625387/213159906-5c29ffe9-34cf-492d-9f57-55738a34b719.png)

**Reply by danielwinkler** [reply]
@garayx thanks a lot for the response. If I understand correctly, then the metadata will be indexed as json string, so there is no way to access the properties inside like:
```
from index 'Companies/TestIndex' where 'Metadata.@collection' == 'Companies'
```

**Reply by garayx** [reply]
you can create a dynamic field from the metadata, and then query it,

```
            Map = companies => from company in companies
                               let metadata = (IEnumerable<KeyValuePair<string, object>>)MetadataFor(company)
                               select new
                               {
                                   Type = company.Type,
                                   ContactsCount = company.Contacts.Count,
                                   LastModified = MetadataFor(company).Value<DateTime>(Constants.Documents.Metadata.LastModified),
                                   _ = metadata.Select(x => CreateField("Metadata_" + x.Key, x.Value, true, true))
                               };
```

query

```
from index 'Companies/CompanyByType' where 'Metadata_@collection' == 'Companies'
```

**Reply by danielwinkler** [reply]
Thanks a lot @garayx!
Very impressive
